cartoframes/datasets.py

```diff
@@ -0,0 +1,252 @@
+import binascii as ba
+from warnings import warn
+
+import pandas as pd
+from carto.exceptions import CartoException
+
+
+class Dataset(object):
+    SUPPORTED_GEOM_COL_NAMES = ['geom', 'the_geom', 'geometry']
+    UTIL_COLS = SUPPORTED_GEOM_COL_NAMES + ['the_geom_webmercator', 'cartodb_id']
+
+    FAIL = 'fail'
+    REPLACE = 'replace'
+    APPEND = 'append'
+
+    def __init__(self, carto_context, table_name, df=None):
+        self.cc = carto_context
```

Comment:
Do we want to have the context in the constructor? The alternative being having to pass it to `write` as a parameter. I'm thinking of local-only usage (without CARTO) e.g: to make a map with VL.

Some options:
- Have a CartoContext and a LocalContext and allow any of them in the constructor? (what if you want to switch it mid-program?)
- Pass the context to operations that write to CARTO like upload (maybe too repetitive to have to pass it everywhere?)
- Have the `write` method be in the context and do `context.write(dataset)`. I think the encapsulation in this is weird.

Reply:
Looking at how Spark works (nothing to do with CARTOframes, but maybe we can do a parallelism), [RDD receive the context in the constructor](http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.rdd.RDD)

So for our case, maybe we can have two contexts (option 1). If you create a Dataset with a LocalContext then that instance is not persistent, otherwise you create a CartoContext and a new Dataset and work with it.

Reply:
I'll discard option #3 since I guess we want users to work with DataFrames not Dataset instances.

About option #2 it depends on what we decide about `CartoContext` if it's the main entry point for CARTOframes then I'll discard it as well.

Reply:
Same as previous one, we should discuss the new API from the user point of view and then think in the implementation

Reply:
I vote in favor of having the context be optional in the constructor. That way local operations are still possible like @javitonino says.
```Dataset(dataset_name, context=None, df=None)```

If a user attempts a write, like: 

```python
>>> d = Dataset('biodiversity_data', df=df)
>>> d.write()
AuthenticanError: pass a context into the write method
>>> d.write(context=cc)
Dataset successfully written to carto
```

So I support option 2. Option 3 is weird but seems natural too. There could be a case where a dataset is associated with one account and you want to send it to another acct.

But overall I agree with @oleurud that it'd be better to do some scoping / looking at this from the user perspective / global view before settling on something.